nearswap/src/twap.rs

```diff
@@ -0,0 +1,553 @@
+            Mean::M_1H => self.to_nanoseconds(60 * 60),
+            Mean::M_12H => self.to_nanoseconds(12 * 60 * 60),
+            _ => 0
+        };
+        let last_index = self.last_updated_index;
+        let req_timestamp = u64::try_from(self.observations[last_index].block_timestamp).unwrap() - time_diff;
+
+        let left_index = self.binary_search(max_length, req_timestamp);
+
+        if left_index == 0 {
+            let total_observe = u128::try_from(self.observations[last_index].num_of_observations).unwrap();
+            let price1cumu = u128::try_from(self.observations[last_index].price1_cumulative).unwrap();
+            let price2cumu = u128::try_from(self.observations[last_index].price2_cumulative).unwrap();
+            let mean1 = price1cumu / total_observe;
+            let mean2 = price2cumu / total_observe;
+            return (U128(mean1), U128(mean2));
+        } else {
+            let total_observe = u128::try_from(self.observations[last_index].num_of_observations).unwrap()
+                                    - u128::try_from(self.observations[left_index - 1].num_of_observations).unwrap();
+            let price1cumu = u128::try_from(self.observations[last_index].price1_cumulative).unwrap()
+                                    - u128::try_from(self.observations[left_index - 1].price1_cumulative).unwrap();
+            let price2cumu = u128::try_from(self.observations[last_index].price2_cumulative).unwrap()
+                                    - u128::try_from(self.observations[left_index - 1].price2_cumulative).unwrap();
+            let mean1 = price1cumu / total_observe;
+            let mean2 = price2cumu / total_observe;
+            return (U128(mean1), U128(mean2));
+        }
+    }
+
+    pub(crate) fn log_observation(&mut self, pool: PoolInfo) {
+        // price1, price2 calculate
+        let near = u128::try_from(pool.ynear).unwrap();
+        let reserve = u128::try_from(pool.reserve).unwrap();
+        let price1: u128 = near / reserve;
+        let price2: u128 = reserve / near;
+
+        // update current index
+        if self.populated == 0 {
+            self.last_updated_index = self.initialize(env::block_timestamp(), price1, price2);
+            self.populated += 1;
+        } else {
+            self.last_updated_index = self.write(
+                env::block_timestamp(),
+                price1,
+                price2,
+                MAX_LENGTH
+            );
+        }
+
+        // update mean
+        let len: usize;
+        if self.pivoted == true {
+            len = MAX_LENGTH.into();
+        } else {
+            len = self.observations.len();
+        }
+
+        self.mean_1min = self.calculate_mean(Mean::M_1MIN, len);
+
+        self.mean_5min = self.calculate_mean(Mean::M_5MIN, len);
+
+        self.mean_1h = self.calculate_mean(Mean::M_1H, len);
+
+        self.mean_12h = self.calculate_mean(Mean::M_12H, len);
+    }
+}
+
+#[derive(Clone, BorshSerialize, BorshDeserialize)]
```

Comment:
We shouldn't need to derive `Clone`, maybe only for tests -> then you should use cfg test attribute:

```
#[cfg_attr(feature = "test", derive(Clone))]
```

Reply:
after `let mut` suggestion, we don't need this, removed clone from Observation